### ISSUE TITLE
Suggestion: check abstract method even if ABCMeta is not used?

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#5026]
It might be really nice if Pyright were to check abstract methods even when ABCMeta is unused.  Currently, inheriting from `ABC` pulls in two generally unwanted things:
- a metaclass `ABCMeta`, which can then prevent inheritance from classes with a different metaclass, and
- a register method on the class for type registration.

Therefore, it's nice to be able to use `abstractmethod` without `ABCMeta`.  MyPy currently checks these abstract methods.  Is there any interest in PyRight doing the same?
```python
from abc import abstractmethod, ABC


class A:  # If ABC is a base class, the error is reported
    @abstractmethod
    def f(self) -> int:
        raise NotADirectoryError


class B(A):
    pass


b = B()  # Would like an error here
```

**Comment by erictraut** [comment]
I've thought about this but decided against it. It would be too computationally intensive. Keep in mind that pyright is a lazy (just-in-time) type evaluator, and it evaluates types of symbols only when needed. Normally, when pyright encounters a constructor call, it evaluates the class declaration (and the class declarations of its base classes, recursively) and the types of the `__new__` and `__init__` methods. It doesn't evaluate the types of any other methods or instance variables because they are not relevant to the constructor call. The exception is if the metaclass for the class hierarchy is `ABCMeta`. In that case, pyright evaluates all of the methods for the entire hierarchy to determine whether any are marked as abstract and are not implemented. This is very expensive, but it's done only for abstract classes. Since the vast majority of classes are not abstract, it's not a big issue. However, with the proposed feature, pyright would need to do this for every class. This would have a big hit on performance and memory usage. That's not a good tradeoff.

If you want pyright to check for missing abstract methods, you should make sure that the base class inherits from `ABC` or a `Protocol`.

**Comment by NeilGirdhar** [comment]
Thanks for explaining!

> when pyright encounters a constructor call, it evaluates the class declaration (and the class declarations of its base classes, recursively) and the types of the __new__ and __init__ methods. 

Right, but you don't need to check if a class has abstract methods at construction, right?  You can do the abstract method check at class declaration time, and memoize that flag.  Then when the class is constructed, you can see if the class is abstract (using the flag), and raise an error if necessary.

It seems like this should only incur a cost based on the number of class declarations (not instantiations), and only cost memory for one flag per class.  Are there really that many class declarations?  I just figured that Pyright does so many other things for a class declaration that one more wouldn't make much of a difference.

> If you want pyright to check for missing abstract methods, you should make sure that the base class inherits from `ABC` or a `Protocol`.

Yeah, unfortunately `ABC` brings in a metaclass, and I'm not implementing a protocol.

**Comment by erictraut** [comment]
>  You can do the abstract method check at class declaration time

The `@abc.abstractmethod` is not part of the class declaration (i.e. the `class` statement); it's a decorator used on individual methods (i.e. `def` statements). When pyright evaluates a class declaration, it doesn't evaluate the types of all of the symbols declared in its body. This is done only on demand if and when a symbol's type is needed.

**Comment by NeilGirdhar** [comment]
> The `@abc.abstractmethod` is not part of the class declaration (i.e. the `class` statement);

Should I have said "definition"?  I mean the entire class body.

> it doesn't evaluate the types of all of the symbols declared in its body. This is done only on demand if and when a symbol's type is needed.

I see.  Thanks for explaining.  Is this redone every time the symbol is used?  Or is the result memoized so that it is only ever done once for a class?

If you do it every time, I guess you're doing that in case the class definition is updated (when an assignment on the class happens, e.g.)?  If so, would a memoization approach be better whereby the memo is invalidated if the class definition is changed?
